extensions/mssql/src/dashboard/bookExtensions.ts

```diff
@@ -0,0 +1,120 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import * as path from 'path';
+import * as arrays from '../util/arrays';
+import { Disposable } from '../util/dispose';
+
+const resolveExtensionResource = (extension: vscode.Extension<any>, resourcePath: string): vscode.Uri => {
+	return vscode.Uri.file(path.join(extension.extensionPath, resourcePath));
+};
+
+const resolveBookResources = (extension: vscode.Extension<any>, books: BookContribution | BookContribution[]): BookContribution[] => {
+	const result: BookContribution[] = [];
+	if (!books) {
+		return result;
+	}
+	if (!Array.isArray(books)) {
+		books = [books];
+	}
+	for (const book of books) {
+		try {
+			book.path = resolveExtensionResource(extension, book.path).fsPath;
+		} catch (e) {
+			// noop
+		}
+		result.push(book);
+	}
+	return result;
+};
+
+export interface BookContribution {
+	name: string;
+	path: string;
+	when?: string;
+}
+
+export namespace BookContributions {
+
+	export function merge(a: BookContribution[], b: BookContribution[]): BookContribution[] {
+		return a.concat(b);
+	}
+
+	export function equal(a: BookContribution, b: BookContribution): boolean {
```

Comment:
typically I think this function is named `equals`

Reply:
I've copied over from markdown extension for all this... so for now going to keep consistent if that's ok.